synapse/axon.py

```diff
@@ -73,24 +70,85 @@ def __init__(self, datadir, **opts):
             self.add()
 
+        url = self.getConfOpt('axonhost:axonbus')
+        if url:
+            self.axonbus = s_service.openurl(url)
+            self.axonbus.runSynSvc(s_common.guid(), self)
+
+    @staticmethod
+    @s_config.confdef(name='axonhost')
+    def _axonhost_confdefs():
+        confdefs = (
+            ('axonhost:autorun', {'type': 'int', 'defval': 0,
+                                  'doc': 'Number of Axons to autostart.'}),
+            ('axonhost:axonbus', {'type': 'str', 'defval': '',
+                                  'doc': 'URL to an axonbus'}),
+            ('axonhost:bytemax', {'type': 'int', 'defval': terabyte,
+                                  'doc': 'Max size of each axon created by the host.'}),
+            ('axonhost:clones', {'type': 'int', 'defval': 2,
+                                 'doc': 'The default number of clones for a axon.'}),
+            ('axonhost:maxsize', {'type': 'int', 'defval': 0,
+                                  'doc': 'Max total allocations for Axons created by the host. '
+                                         'Only applies if set to a positive integer.'}),
+            ('axonhost:syncmax', {'type': 'int', 'defval': gigabyte * 10,
+                                  'doc': ''}),  # XXX Words
+            ('axonhost:hostname', {'type': 'str', 'defval': s_thishost.get('hostname'),
+                                   'doc': 'AxonHost hostname'}),
+        )
+        return confdefs
+
     def _fireAxonIden(self, iden):
         axondir = s_common.gendir(self.datadir, '%s.axon' % iden)
 
-        opts = dict(self.opts)
+        opts = self.makeAxonOpts()
         jsopts = s_common.jsload(axondir, 'axon.opts')
         if jsopts is not None:
             opts.update(jsopts)
 
+        # Special case where the axonbus may update - we want to ensure
+        # we're passing the latest axonbus to the Axon so it can register
+        # itself properly.
+        if 'axon:axonbus' in opts:
+            axonbus = opts.get('axon:axonbus')
+            myaxonbus = self.getConfOpt('axonhost:axonbus')
+            if axonbus != myaxonbus:
+                opts['axon:axonbus'] = myaxonbus
+                s_common.jssave(opts, axondir, 'axon.opts')
+
         self.axons[iden] = Axon(axondir, **opts)
 
+        bytemax = opts.get('axon:bytemax')
+        clone = opts.get('axon:clone')
+
+        if clone:
+            self.cloneaxons.append(iden)
+        else:
+            bytemax += opts.get('axon:syncmax')
+        self.usedspace = self.usedspace + bytemax
+
     def _onAxonHostFini(self):
         for axon in list(self.axons.values()):
             axon.fini()
 
+    def makeAxonOpts(self):
+        '''
+        Make a configable dictionary from the current AxonHost options.
+
+        Returns:
+            dict: Configable dict of valid options which can be passed to a Axon()
+        '''
+        ret = {}
+        for name, _ in self.getConfDefs().items():
```

Comment:
we could avoid this if we made the axonhost config options that were specifically for individual axons start with just axon: and get passed through.  They are essentially the "axon options known by axonhost" so it seems clean.

Reply:
This is now cleaner but we still have to do some examination of the options since we can't just transparently pass options down to the axon.